xwiki-platform-core/xwiki-platform-refactoring/xwiki-platform-refactoring-default/src/main/java/org/xwiki/refactoring/internal/XDOMReferenceRenamer.java

```diff
@@ -0,0 +1,207 @@
+import org.xwiki.rendering.block.match.BlockMatcher;
+import org.xwiki.rendering.block.match.ClassBlockMatcher;
+import org.xwiki.rendering.block.match.OrBlockMatcher;
+import org.xwiki.rendering.listener.reference.ResourceReference;
+import org.xwiki.rendering.listener.reference.ResourceType;
+
+/**
+ * Internal component used to replace easily references in blocks.
+ *
+ * @version $Id$
+ * @since 13.4RC1
+ */
+@Singleton
+@Component(roles = XDOMReferenceRenamer.class)
+public class XDOMReferenceRenamer
+{
+    private static final List<BlockMatcher> DEFAULT_BLOCK_MATCHERS = Arrays.asList(
+        new ClassBlockMatcher(LinkBlock.class),
+        new ClassBlockMatcher(ImageBlock.class)
+    );
+
+    private static final Set<ResourceType> SUPPORTED_RESOURCE_TYPES = new HashSet<>(Arrays.asList(
+        ResourceType.DOCUMENT,
+        ResourceType.SPACE,
+        ResourceType.PAGE,
+        ResourceType.ATTACHMENT
+    ));
+
+    @Inject
+    private EntityReferenceResolver<ResourceReference> entityReferenceResolver;
+
+    @Inject
+    @Named("compact")
+    private EntityReferenceSerializer<String> compactEntityReferenceSerializer;
+
+    @Inject
+    private DocumentReferenceResolver<EntityReference> defaultReferenceDocumentReferenceResolver;
+
+    @Inject
+    private PageReferenceResolver<EntityReference> defaultReferencePageReferenceResolver;
+
+    private static final String DEFAULT_SPACE_HOMEPAGE = "WebHome";
+
+    /**
+     * @param xdom the XDOM to extract from
+     * @return the list of blocks that link to documents or spaces.
+     */
+    public List<Block> getBlocks(XDOM xdom, boolean withMacroBlock)
+    {
+        List<BlockMatcher> blockMatchers;
+        if (withMacroBlock) {
+            blockMatchers = new ArrayList<>(DEFAULT_BLOCK_MATCHERS);
+            blockMatchers.add(new ClassBlockMatcher(MacroBlock.class));
+        } else {
+            blockMatchers = DEFAULT_BLOCK_MATCHERS;
+        }
+        return xdom.getBlocks(new OrBlockMatcher(blockMatchers), Block.Axes.DESCENDANT);
+    }
+
+    public boolean renameReferenceInBlock(Block block, DocumentReference oldReference, DocumentReference newReference,
+        DocumentReference currentDocumentReference, boolean relative)
+    {
+        ResourceReference reference;
+        if (block instanceof LinkBlock) {
+            LinkBlock linkBlock = (LinkBlock) block;
+            reference = linkBlock.getReference();
+
+        } else if (block instanceof ImageBlock) {
+            ImageBlock imageBlock = (ImageBlock) block;
+            reference = imageBlock.getReference();
+        } else {
+            throw new IllegalArgumentException(String.format(
+                "Only LinkBlock and ImageBlock can be processed in XDOMReferenceRenamer. Given class was: [%s]",
+                block.getClass().getName()));
+        }
+
+        if (reference == null) {
+            throw new IllegalArgumentException("The reference of  the block cannot be null.");
```

Comment:
Two spaces between `of` and `the`.

Reply:
And always good ot put the max info in the log or exception message. The reference should probably be indicated here.

Reply:
I can improve it yes, note that this check is coming from the old ReferenceRenamer which was only throwing an empty IllegalArgumentException back then: I'm even wondering if this could really happen...